_drafts/2024-05-18-get-started-with-aws-tf-01.md

```diff
@@ -0,0 +1,254 @@
+1. An AWS EC2 instance resource
+2. An aws ami image data reference for EC2 to run
+
+Create a new file called `unicorn-api.tf`.  
+Inside the file, add:
+
+```terraform
+# /infra/unicorn-api.tf
+
+data "aws_ami" "amazon_x86_image" {
+    most_recent = true
+    owners = ["amazon"]
+    name_regex = "^al2023-ami-ecs.*"
+
+    filter {
+        name = "architecure"
+        values = ["x86_64"]
+    }
+}
+
+resource "aws_instance" "unicorn_vm" {
+    ami = data.aws_ami.amazon_x86_image.id
+    instance_type = "t2.micro"
+
+    tags = {
+        Name = "Unicorn"
+    }
+}
+```
+
+Now when you open the terminal and run `terraform plan` the logs will show that we're adding one new resource.  
+
+Now try running `terraform apply` to... err... apply the plan.  
+
+It fails! But why?
+
+Even though we've created an aws account, and downloaded Terraform, and the aws provider for Terraform to use, we still need Terraform to authenticate with aws in order to create these resources.
+
+## Step 2.1 - Create an AWS API Key for terraform
+
+In order for Terraform to manage resources in AWS it needs access to an aws account. We can accomplish this by creating an aws account specifically for this tool and defining corresponding API keys for it.
+
+Create a new file called `aws-provider-key.tf` and add the following content:
+
+```terraform
+# /infra/aws-provider-key.tf
+
+provider "aws" {
+    access_key = ""
+    secret_key = ""
+    region = "ap-southeast-2"
+}
+```
+
+You can opt to use a region closer to you. As I'm in Australia I'm opting to use the australian region.
+
+Next you need to create two things in the AWS Management Console:
+1. Create a [new User](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html) with administrator permission
+2. Create an [access key](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html) for the user
+
+Take the access key and corresponding secret key and add them to your provider key your defined earlier.
+
+> These API keys should be treated like credentials. Do not commit this file to source control. Otherwise the key could be discovered and you'll need to replace them.
+
+Now try running `terraform apply` once more and everything should be hunky dory.
+
+## Step 3 - Connect to our EC2 Instance
+
+If you've made it this far it means you've successfully configured terraform with an aws account and have deployed an EC2 instance to AWS. Nice!  
+Let's try connect to this instance.  
+We can find the ip address by logging in to the AWS Management Console and opening the instance details within EC2.
+Copy the IPv4 address and in a terminal run `ping <the-ip-address>`.
+
+No response...  
+
+This is because when we define the EC2 instance, a default security group is applied to it. This security group allows incoming connections on any port using any protocol but only from other aws resources with the same security group rules...
```

Comment:
```suggestion
This is because when we define the EC2 instance, a default security group is applied to it. This security group allows incoming connections on any port using any protocol but only from other aws resources with the same security group...
```